Add XmlProcessor initial implementation

marc-gr opened the pull request:
This PR creates a new XML processor that achieves feature parity with Logstash's XML filter.

### ⚙️ **Configuration Options**

```yaml
processors:
  - xml:
      field: "xml_data"
      target_field: "parsed"
      to_lower: false
      # Logstash-compatible options
      xpath:
        "/root/item/@id": "item_id"
        "//product/name/text()": "product_name"
      namespaces:
        "ns": "http://example.com/namespace"
      force_array: true
      force_content: false
      remove_namespaces: false
      ignore_empty_value: true
      parse_options: "strict"
```

### 🏗️ **Architecture**

- **Streaming SAX Parser**: Optimal memory usage for large XML documents
- **Selective DOM Building**: Only builds DOM when XPath expressions are configured
- **Pre-compiled XPath**: XPath expressions compiled at processor creation for performance
- **Security**: Enhanced XXE protection with secure parser factory configurations


### 📚 **Documentation**

Documentation includes:
- Complete configuration reference
- XPath expression examples
- Namespace configuration guide

### **Logstash differences**

- `ignore_empty_value` behaves a bit different than `suppress_empty`, but I think it matches better with other processors behavior. It could be adapted, or even add both, but I found it confusing.

Closes https://github.com/elastic/elasticsearch/issues/97364

marc-gr committed:
Add XmlProcessor initial implementation

Commit:
[CI] Auto commit changes from spotless

marc-gr committed:
Make factory static

Commit:
[CI] Auto commit changes from spotless

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

marc-gr committed:
Update docs/changelog/130337.yaml

marc-gr committed:
Merge remote-tracking branch 'upstream/main' into feat/xml-processor

marc-gr committed:
feat: rewrite XML processor for Logstash feature parity

- Replace XMLStreamReader with SAX parser + DOM for XPath support
- Add XPath extraction, namespaces, strict parsing, content filtering
- New options: force_array, force_content, remove_namespaces, store_xml
- Enhanced security with XXE protection and pre-compiled XPath expressions
- Full test coverage and updated documentation

Commit:
[CI] Auto commit changes from spotless

marc-gr committed:
fix: implement Copilot PR review suggestions

- Fix test assertion for remove_namespaces feature
- Use StandardCharsets.UTF_8 instead of string literal
- Replace string reference comparison with isEmpty()
- Move regex pattern to static final field for performance

marc-gr committed:
Merge branch 'feat/xml-processor' of github.com:marc-gr/elasticsearch into feat/xml-processor

Commit:
[CI] Auto commit changes from spotless

marc-gr committed:
Merge remote-tracking branch 'upstream/main' into feat/xml-processor

marc-gr committed:
Merge branch 'feat/xml-processor' of github.com:marc-gr/elasticsearch into feat/xml-processor

marc-gr committed:
Merge branch 'main' into feat/xml-processor

marc-gr committed:
Merge branch 'main' into feat/xml-processor

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
```diff
@@ -0,0 +1,575 @@
+          "foo": {
+            "b": "bold",
+            "i": "italic",
+            "#text": "This text is  and this is !"
```
For what it's worth, I was somewhat surprised at this is behaviour. The text and the tags are all there, but the information about how they are ordered and interleaved has been thrown away.

You're the one who has actual use-cases, and you know how logstash behaves, so if you tell me that this is a sensible thing to do then I'll believe you. I just wanted to check that it's deliberate.

PeteGillinElastic replied:
Okay, thanks!

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
```diff
@@ -0,0 +1,575 @@
+| `ignore_missing` | no | `false` | If `true` and `field` does not exist, the processor quietly exits without modifying the document. |
+| `ignore_failure` | no | `false` | Ignore failures for the processor. When `true` and XML parsing fails, adds `_xmlparsefailure` tag to the document. See [Handling pipeline failures](docs-content://manage-data/ingest/transform-enrich/ingest-pipelines.md#handling-pipeline-failures). |
+| `to_lower` | no | `false` | Convert XML element names and attribute names to lowercase. |
+| `ignore_empty_value` | no | `false` | If `true`, the processor will filter out null and empty values from the parsed XML structure, including empty elements, elements with null values, and elements with whitespace-only content. |
```
Could we consider maybe calling this `remove_empty_value`? Or even `remove_empty_values`? It seems to me that this would make it seem logically related to `remove_namespaces`, rather than to `ignore_missing` and `ignore_failure`, which more accurately reflects what it does.

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
```diff
@@ -0,0 +1,575 @@
+| `remove_namespaces` | no | `false` | If `true`, removes namespace prefixes from element and attribute names. |
+| `force_content` | no | `false` | If `true`, forces text content and attributes to always parse to a hash value with `#text` key for content. |
+| `force_array` | no | `false` | If `true`, forces all parsed values to be arrays. Single elements are wrapped in arrays. |
+| `parse_options` | no | - | Controls XML parsing behavior. Set to `"strict"` for strict XML validation that fails fast on invalid content. |
```
Is there a reason to make this a string-valued option where `"strict"` is the only valid value, rather than e.g. a boolean `strict_parsing` option?

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
````diff
@@ -0,0 +1,575 @@
+}
+```
+
+### Converting element names to lowercase
````
It's up to you, but in the interests of brevity, I'd consider omitting this section. It seems pretty obvious from the description above what this option does, I can't really see anyone needing the example.

But if you want this here e.g. to draw extra prominence to the option, it's okay to leave it.

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
```diff
@@ -0,0 +1,575 @@
+| --- | --- | --- | --- |
+| `field` | yes | - | The field containing the XML string to be parsed. |
+| `target_field` | no | `field` | The field that the converted structured object will be written into. Any existing content in this field will be overwritten. |
+| `store_xml` | no | `true` | If `true`, stores the parsed XML structure in the target field. If `false`, only XPath extraction results are stored. |
```
Am I correct in figuring that `target_field` is ignored if this is set? I think it might be worth stating that explicitly.

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
````diff
@@ -0,0 +1,575 @@
+    }
+  ]
+}
+```
````
Please forgive my ignorance, it's a long time since I've dealt with XML namespaces (or XML at all, to be honest!). Why do we need to include the namespace mappings in the processor definition, when — in this example, at least — the same mappings are provided via the `xmlns:*` attributes in the source document?

PeteGillinElastic replied:
Yeah, that makes sense now that you explain it.

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
````diff
@@ -0,0 +1,575 @@
+}
+```
+
+### Force array behavior
````
I sort of wonder whether this section adds much value, although I feel less strongly about this than about the lower case one.

PeteGillinElastic commented on docs/reference/enrich-processor/xml-processor.md:
```diff
@@ -0,0 +1,575 @@
+  "docs": [
+    {
+      "_source": {
+        "xml_content": "<catalog><book><title>Invalid XML with control character</title></book></catalog>"
```
Perhaps I'm just being blind here, but... I don't see any control character here?

PeteGillinElastic reviewed (commented):
Hi @marc-gr. First of all, thanks again for the contribution.

I figured that a sensible way to approach the review here is to cover the behaviour in a first round, and then cover the implementation later. Since you helpfully included comprehensive documentation, I figured that would be a good way into it. As such, in this review I've looked at those files, but I haven't looked at the actual Java code at all. I'll do that in a later round.

Overall, I think that the behaviour looks sensible, and the docs are nice and clear. I've left a few comments, which are mostly presentational, or asking for clarification.

marc-gr committed:
Sort processor reference

marc-gr committed:
Better doc for force_content

marc-gr committed:
Better changelog entry

marc-gr committed:
Rename ignore_empty_value to remove_empty_values

marc-gr committed:
Add clarification in docs for store_xml

marc-gr committed:
Change strict parsing option to be boolean

marc-gr committed:
Remove unnecessary docs

marc-gr committed:
Merge remote-tracking branch 'upstream/main' into feat/xml-processor

marc-gr committed:
Merge branch 'feat/xml-processor' of github.com:marc-gr/elasticsearch into feat/xml-processor

marc-gr committed:
Fix style checks

PeteGillinElastic reviewed (commented):
Hi. Sorry it's taken me a while to get to this... There are quite a few comments here, although hopefully not an unreasonable number given that it's nearly 1000 lines of code! Most of them are stylistic nits (and a bunch of them are repeats). There are a couple of slightly more substantive questions that I think we should address. But the overall shape of the code looks fine.

I haven't looked at the tests yet, I'll try to get to those on the next round.

marc-gr committed:
Improve XmlProcessor code quality and exception handling

- Use pattern matching for instanceof in XPath and namespace configuration parsing
- Replace manual null/whitespace checks with Strings.hasText() utility method
- Clean up exception messages by removing redundant e.getMessage() concatenation
- Add ByteArrayInputStream import instead of fully qualifying at point of use
- Fix Javadoc formatting to use proper HTML lists instead of markdown syntax
- Update parameter names in processXPathExpressionsFromDom for clarity
- Simplify exception handling by letting XPathExpressionException bubble up
- Update tests to match cleaner exception message format

marc-gr commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,964 @@
+
+    /**
+     * Creates a secure, pre-configured SAX parser factory for XML parsing.
+     * This factory is configured to prevent XXE attacks with SAX-specific features.
```
Good point, it'd be good to have someone take a look if you know the person. I went through with docs and what gemini suggested, since I am far from an expert in this regard

PeteGillinElastic replied:
I'll see if I can find a good person.

marc-gr committed:
Refactor XmlProcessor configuration handling and enhance factory tests for validation

Commit:
[CI] Auto commit changes from spotless

PeteGillinElastic commented:
Thanks. I've looked at all the latest changes and everything seems good. I haven't had time to review the unit tests yet, sorry, I'll get to that next week (please paster me if it looks like I'm not keeping that promise!). And, as mentioned, I'll try to see if I can find someone appropriate to look at the parser security stuff.

PeteGillinElastic commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/XmlProcessorTests.java:
```diff
@@ -0,0 +1,638 @@
+
+        // Get the XPath result
+        Object barContent = ingestDocument.getFieldValue("bar_content", Object.class);
+        assertNotNull(barContent);
```
No need for this. The next assertion will fail if it's null, and will give the reader more information about what we actually expected to find.

I'm not sure that _any_ of the calls to `assertNotNull` in this class are needed.

PeteGillinElastic commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/XmlProcessorTests.java:
```diff
@@ -0,0 +1,638 @@
+        // Get the XPath result
+        Object barContent = ingestDocument.getFieldValue("bar_content", Object.class);
+        assertNotNull(barContent);
+        assertEquals("hello", barContent);
```
This is kind of pedantic, but we're mixing junit and hamcrest assertions here... It's mildly better to stick to just hamcrest, i.e. `assertThat(barContent, equalTo("hello"))`. It's a bit confusing having the order of the actual and expected parameters switching around like this.

PeteGillinElastic commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/XmlProcessorTests.java:
```diff
@@ -0,0 +1,638 @@
+        String xml = "<root>"
+            + "  <person id=\"1\"><n>John</n><age>30</age></person>"
+            + "  <person id=\"2\"><n>Jane</n><age>25</age></person>"
+            + "</root>";
```
For these strings which don't fit on one line, it's neater to do this:
```java
String xml = """
    <root>
      <person id="1"><n>John</n><age>30</age></person>
      <person id="2"><n>Jane</n><age>25</age></person>
    </root>""";
```

This isn't exactly the same as what you have, because my string includes newlines. You can put a `\` at the end of the line if you want to omit the newline, but I think it's fine to leave it in.

In fact, you may prefer something like this:
```java
String xml = """
    <root>
      <person id="1">
        <n>John</n>
        <age>30</age>
       </person>
      <person id="2">
        <n>Jane</n>
        <age>25</age>
       </person>
    </root>""";
```

so that the structure is more obvious to the reader. Since the text-block syntax makes breaking across lines so easy, we might as well format the XML like we normally would when pretty-printing.

PeteGillinElastic replied:
Same applies throughout this class. Optionally, you could even do this for the strings which _do_ fit in one line, so that you can pretty-print.

PeteGillinElastic commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/XmlProcessorTests.java:
```diff
@@ -0,0 +1,638 @@
+
+        assertTrue(
+            "Error message should indicate XML is invalid",
+            exception.getMessage().contains("invalid XML") || exception.getCause().getMessage().contains("XML")
```
Can we really not assert which of the two things is true?

PeteGillinElastic commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/XmlProcessorTests.java:
```diff
@@ -0,0 +1,638 @@
+        // Verify field names are lowercase
+        Map<String, Object> data = ingestDocument.getFieldValue(TARGET_FIELD, Map.class);
+        assertTrue(data.containsKey("foo"));
+        assertFalse(data.containsKey("FOO"));
```
The hamcrest forms would be
```java
assertThat(data, hasKey("foo"));
assertThat(data, not(hasKey("FOO"));
```
though, again, we could consider a stronger assertion here.

PeteGillinElastic replied:
Same thing applies in other places.

marc-gr committed:
Merge branch 'main' of github.com:elastic/elasticsearch into feat/xml-processor

marc-gr committed:
Simplify tests and make them more idiomatic

marc-gr committed:
Use provided secured xml factories

marc-gr committed:
Merge branch 'main' into feat/xml-processor

marc-gr committed:
Merge branch 'feat/xml-processor' of github.com:marc-gr/elasticsearch into feat/xml-processor

Commit:
[CI] Auto commit changes from spotless

PeteGillinElastic commented:
> ## ℹ️ Important: Docs version tagging
> 👋 Thanks for updating the docs! Just a friendly reminder that our docs are now **cumulative**. This means all 9.x versions are documented on the same page and published off of the main branch, instead of creating separate pages for each minor version.
> 
> We use [applies_to tags](https://elastic.github.io/docs-builder/syntax/applies) to mark version-specific features and changes.
> 
> Expand for a quick overview
> ### When to use applies_to tags:
> ✅ At the page level to indicate which products/deployments the content applies to (mandatory) ✅ When features change state (e.g. preview, ga) in a specific version ✅ When availability differs across deployments and environments
> 
> ### What NOT to do:
> ❌ Don't remove or replace information that applies to an older version ❌ Don't add new information that applies to a specific version without an applies_to tag ❌ Don't forget that applies_to tags can be used at the page, section, and inline level
> 
> ### 🤔 Need help?
> * Check out the [cumulative docs guidelines](https://elastic.github.io/docs-builder/contribute/cumulative-docs/)
> * Reach out in the [#docs](https://elastic.slack.com/archives/C0JF80CJZ) Slack channel

@marc-gr I think we need to act on this, thanks to the new way that docs work in ES. My reading of the [docs docs](https://elastic.github.io/docs-builder/syntax/applies/) is that we don't have to do anything for serverless (the feature is GA and serverless is unversioned) but that we will need to declare something like `stack: ga 9.2`. I think you'll need to do it at the page level in the new page, and inline for the index change (I don't think it's needed for the `toc.yml`.) It looks like you should be able to use the `normalize-for-stream` processor as an example.

(Thanks for your latest changes, I'll try to have a look in the next day or two.)

marc-gr committed:
Merge branch 'main' into feat/xml-processor

marc-gr committed:
Add applies_to section for XML processor documentation

marc-gr committed:
Merge branch 'feat/xml-processor' of github.com:marc-gr/elasticsearch into feat/xml-processor

marc-gr committed:
Merge branch 'main' into feat/xml-processor

marc-gr committed:
Merge branch 'main' into feat/xml-processor

marc-gr committed:
Merge branch 'main' into feat/xml-processor

PeteGillinElastic commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,915 @@
+    private static final SAXParserFactory SAX_PARSER_FACTORY_NS_STRICT = createSecureSaxParserFactoryNamespaceAwareStrict();
+
+    // Pre-configured secure document builder factory for DOM creation
+    private static final DocumentBuilderFactory DOM_FACTORY = createSecureDocumentBuilderFactory();
```
Sorry, I should have mentioned this in a previous review round, it slipped my mind... 

These `createSecure*` methods will throw if the user's JDK doesn't provide the required functionality. Because we call them from static initializers, this will happen when the class is loaded by the JVM, which (because of the way we register ingest processors) will be during node construction. So the node will fail to start, even if they don't use this functionality. That seems like something we should try to avoid. 

I propose making two changes here.

First change: In the catch block in each the `createSecure*` method, instead of propagating the exception, log it and return null; then in `selectSaxParserFactory()` do check if the selected factory is null and throw `UnsupportedOperationException` with a nice error message if so. That way, we only fail when the functionality is actually used.

The remaining problem is that we'll log the original exception during node construction. This happens before we know whether the functionality will be used or not, and at a time when a _lot_ of logging is happening and it could easily be missed, rather than when the functionality is used.

So, second change: Move these five constants into a nested inner class e.g. `private static class XmlFactories { ... }`. That inner class will only get loaded, and so the exceptions will get logged, the first time the functionality is used. This is basically a handy way of doing a lazily-initialized singleton pattern.

(The second change by itself would actually fix the basic issue of the nodes failing to start up, but the exceptions you'd get out of it would be a lot less helpful, so I think making both changes is the most user-friendly approach.)

(All this is somewhat theoretical: Our bundled JDK obviously does provide this stuff, and I would be surprised if any of the JDKs we support don't. However, it seems worth making a small effort to behave as nicely as possible if the user is doing something funky.)

PeteGillinElastic commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,915 @@
+            try {
+                factory.setFeature("http://apache.org/xml/features/validation/check-full-element-content", true);
+            } catch (Exception e) {
+                // Strict parsing features are optional - continue without them if not supported
```
So... If the user has set `strict_parsing`, and it's not supported, is silently ignoring it definitely the right thing to do? I kinda feel it's not.

At the very least, I think we should log.warn the exception and its significance (i.e. that we're going to ignore strict_parsing).

However, might it not be better to actually fail? Then the user can decide whether they'd like to do without `strict_parsing`, or whether they'd like to fix whatever issue with their JDK means it's not supported, or to find some other solution. If you go with my advice from the comment above, I think you can simply omit this inner try-catch block: if, for some reason, the JDK supports the other options but not this one, then the other factories will be initialized and this one will be null and the user will just get an exception if they've set `strict_parsing`. (We should make sure that the exception we throw tells the user that such-and-such a combination of features isn't working, so they can try others.)

PeteGillinElastic commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,915 @@
+            try {
+                factory.setFeature("http://apache.org/xml/features/validation/check-full-element-content", true);
+            } catch (Exception e) {
+                // Strict parsing features are optional - continue without them if not supported
```
Same comment as above.

PeteGillinElastic commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,915 @@
+        // Use the appropriate pre-configured SAX parser factory
+        SAXParserFactory factory = selectSaxParserFactory();
+
+        SAXParser parser = factory.newSAXParser();
```
Another thing which should probably have occurred to me before: I have a nasty feeling that creating a new parser for every document is pretty inefficient. There are a few options here. I'd like to get a second opinion from the team, so I'll start a discussion (and I'll @-mention and notify you).

PeteGillinElastic commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/XmlProcessorTests.java:
```diff
@@ -0,0 +1,605 @@
+/**
+ * Tests for {@link XmlProcessor}. These tests ensure feature parity and test coverage.
+ */
+@SuppressWarnings("unchecked")
```
Do you still need this? At a glance, I didn't see any unchecked casts left, but it's a long file and I could easily have missed them!

If there's a small number, consider suppressing at a narrower scope, i.e. the method or assignment statement where it happens.

PeteGillinElastic reviewed (commented):
Hi. As it happens, I've been slowly working through the latest review round... Thanks for all this!

The docs changes look good. I checked the previews and it has put `Stack | Planned` in the appropriate places, which makes sense since 9.2 hasn't been released yet (I assume that it will magically update when it has).

The test changes look good. Unfortunately, a couple of other things occurred to me as I was looking through it again. I've left comments. Sorry this stuff didn't occur to me before. Mostly I think they should be pretty quick to implement. As mentioned, there's one issue I want to follow up on elsewhere.

marc-gr commented:
/cla

marc-gr committed:
Refactor XmlProcessor to use lazy-initialized XML factories and add logging for configuration issues; update tests to suppress unchecked warnings.

marc-gr commented:
I think all is addressed 👍

Commit:
[CI] Auto commit changes from spotless

PeteGillinElastic commented:
Thanks. Those last changes look good, with a couple of tiny nits.

I think there are two outstanding items here:
 1. The performance questions which @samxbr is going to help with.
 2. The fact that turning on strict mode causes exceptions in CI, which we have a thread about elsewhere.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,967 @@
+        String xmlString = (String) fieldValue;
+        try {
+            // Always use streaming parser for optimal performance and memory usage
+            if (storeXml || xpathExpressions.isEmpty() == false) {
```
Why do this `if` have to be in a `try`/`catch`? It doesn't seem relevant to the exception handling. If the goal is early return in certain scenarios, then write that code simply. This is very confusing as written.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,942 @@
+        }
+
+        // Determine if we need DOM for XPath processing
+        boolean needsDom = xpathExpressions.isEmpty() == false;
```
I don't think you need to do work here on a per-document basis. Do you?

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,942 @@
+        boolean needsDom = xpathExpressions.isEmpty() == false;
+
+        // Use the appropriate pre-configured SAX parser factory
+        SAXParserFactory factory = selectSaxParserFactory();
```
I don't think this needs to be done per-document either. My casual read of it is that the parser factory could be a final variable selected as configuration parse time (that is, at processor construction time).

For this and my previous comment, it's less about **performance** than it is about making the code obvious.

Things that do not **have** to be executed on a **per-document** basis shouldn't be.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,942 @@
+            Object result = compiledExpression.evaluate(xmlDocument, XPathConstants.NODESET);
+
+            if (result instanceof NodeList nodeList) {
+                List<String> values = new ArrayList<>();
```
It might be clearer if you split the code into the 0-1 case (where nodeList.getLength() <= 1), and where you do not have to allocate an `ArrayList` at all, and the 0-N case (where nodeList.getLength() > 1), and where you do need to allocate an `ArrayList` (but _might_ still end up only having a single result).

This is partially a code clarity point and partially a performance point. (I haven't benchmarked this, but avoiding  potentially many un-necessary ArrayList allocations per document is almost certainly a win at the margins.)

joegallo replied:
I think @samxbr handled this comment via 998cae6b4ba7dfc0246f947118a17dbabedbff1d.

marc-gr committed:
Remove strict_parsing option

samxbr commented:
@joegallo has readily added the ThreadLocal version parser in [his branch](https://github.com/joegallo/elasticsearch/commits/feat/xml-processor/), and doing the [same benchmark](https://github.com/elastic/elasticsearch/pull/130337#issuecomment-3272849327) showed a much better processor latency of `~2.6us` (from `~15 us`):
```
"xml":{"count":10000005,"failed":0,"current":0,"time_in_millis":26762}}}
```

samxbr committed:
Joe's thread local and soft reference

samxbr committed:
Merge branch 'main' into feat/xml-processor

samxbr committed:
address comments and some minor refactoring

joegallo committed:
Merge branch 'main' into feat/xml-processor

joegallo committed:
Silence some warnings from IntelliJ

joegallo committed:
Silence some warnings from IntelliJ

joegallo committed:
Silence some warnings from IntelliJ

joegallo committed:
Silence some warnings from IntelliJ

joegallo committed:
Prefer imports over qualified names

joegallo committed:
This can be static

joegallo committed:
Read from the String directly

joegallo committed:
Don't use Hungarian notation

joegallo committed:
Clarify logic and add a TODO

joegallo committed:
Handle input more conventionally

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,891 @@
+            throw new IllegalArgumentException("field [" + field + "] is null, cannot parse XML");
+        }
+
+        // TODO this just seems like misconfiguration to me. Why should we all this at all?
```
/cc @PeteGillinElastic since you brought this up earlier in https://github.com/elastic/elasticsearch/pull/130337#discussion_r2291947785.

joegallo committed:
Rewrite all these conditionals

There was a warning here about hasChildren always being true at one
point, and it just seemed like the overall logic was easier to follow
as nested ifs (to me).

joegallo committed:
Avoid an unnecessary containsKey

joegallo committed:
Tighten up the collections handling here

This now results in an immutable map around immutable sets, with no
unmodifiable wrappers.

PeteGillinElastic commented:
Sorry, github has been being weird at me and sticking some of my comments in pending when I didn't expect it to. I was meaning to reply to the question about the noop processor. The other comments are nits, I'm not sure whether they're still valid on the latest iteration of the code.

joegallo committed:
Merge branch 'main' into feat/xml-processor

joegallo committed:
Invert these conditionals for consistency reasons

joegallo committed:
Always use the utility function

joegallo committed:
This can be static

joegallo committed:
This doesn't need to handle the general case

The callers are always either static-ly a String or a List, so let's
just have static arities for exactly and only those versions.

joegallo committed:
Whitespace

They don't make a drug for what's wrong with me.

joegallo committed:
Only trim the text if it's not blank

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,879 @@
+        this.xpathExpressions = xpathExpressions != null ? Map.copyOf(xpathExpressions) : Map.of();
+        this.namespaces = namespaces != null ? Map.copyOf(namespaces) : Map.of();
+
+        this.compiledXPathExpressions = compileXPathExpressions(this.xpathExpressions, this.namespaces);
```
From https://jcp.org/aboutJava/communityprocess/review/jsr063/jaxp-pd1.pdf, though, it seems that the `DocumentBuilderFactory` and `SAXParserFactory` classes are threadsafe, so `XPathFactory` is the odd ball. This is fun, we're having fun.

joegallo committed:
Merge branch 'main' into feat/xml-processor

joegallo commented:
I merged `main` in to take advantage of https://github.com/elastic/elasticsearch/pull/134923, but I haven't addressed my complaints about the other aspects of the XPath handling here yet.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java:
```diff
@@ -0,0 +1,889 @@
+            return XmlUtils.getHardenedXPath();
+        } catch (Exception e) {
+            logger.warn("Cannot configure secure XPath object - XML processor may not work correctly", e);
+            return null;
```
I followed the existing pattern of returning null here, but honestly I think we should blow up and die much more aggressively if this happens. The processor is just going to throw NPEs at every step of the way anyway, and that seems... unproductive.

joegallo committed:
Handle these options in the opposite order

joegallo committed:
Use better variable names here

joegallo committed:
Handle this validation in the factory

joegallo committed:
Hoist the namespace context into a field

joegallo committed:
Update this test string

Commit:
[CI] Update transport version definitions

Commit:
[CI] Update transport version definitions
